frontends/p4/typeChecking/typeChecker.cpp

```diff
@@ -4149,6 +4153,28 @@ const IR::Node *TypeInference::postorder(IR::AssignmentStatement *assign) {
     return assign;
 }
 
+const IR::Node *TypeInference::postorder(IR::ForInStatement *forin) {
+    LOG3("TI Visiting " << dbp(getOriginal()));
+    auto ltype = getType(forin->ref);
+    if (ltype == nullptr) return forin;
+
+    if (!isLeftValue(forin->ref)) {
+        typeError("Expression %1% cannot be the target of an assignment", forin->ref);
+        LOG2(forin->ref);
+        return forin;
+    }
+    if (auto range = forin->collection->to<IR::Range>()) {
```

Comment:
Question: Is it possible for `collection` to not be `IR::Range`?

Reply:
It seems so: https://github.com/p4lang/p4c/pull/4562/files#diff-85fbe45d422156081940f1eb7e38cd4ff1bd03574da101de1cf7021a166851c9R1349

Reply:
Then is it possible to also check the type for non-range `collection`s?

Reply:
Yes, there should be some other checks here for that case, but I'm still unsure as to exactly the full scope of iterations over 'collections'.  Need support for `list<T>` and `header_stack` at least; maybe other things as well?

Reply:
maybe we can do the opposite and explicitly require a `Range` here? lists and header stacks are so different conceptually here